### ISSUE TITLE
Overriding MutableMapping.setdefault() doesn seem to process optional argument values

### ISSUE THREAD

**Issue by slott56** [microsoft/pyright#7362]
**Describe the bug**
An override of `MutableMapping` (or `dict`) that attempts to define `setdefault()` reports a specious error around the default value positional parameter.

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
class MD(dict[str, str]):
    def setdefault(self, key: str, default: str | None = None) -> str:
        return super().setdefault(key, default)  # Line 3

class MDO(dict[str, str]):
    def setdefault(self, key: str, default: str | None = None) -> str:
        if default is None:
            return super().setdefault(key)  # Line 8
        else:
            return super().setdefault(key, default)

if __name__ == "__main__":
    md = MD(hello="World!")
    assert md.setdefault('one') is None
    assert md.setdefault('two', 'more') == 'more'
    assert md == {'hello': 'World!', 'one': None, 'two': 'more'}
```

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
This is run on the command-line

```bash
% pyright spike
/Users/slott/Documents/Projects/ChatTUI/spike/bug_report.py
  /Users/slott/Documents/Projects/ChatTUI/spike/bug_report.py:3:40 - error: Argument of type "str | None" cannot be assigned to parameter "__default" of type "str" in function "setdefault"
    Type "str | None" cannot be assigned to type "str"
      "None" is incompatible with "str" (reportArgumentType)
  /Users/slott/Documents/Projects/ChatTUI/spike/bug_report.py:8:20 - error: Expected 1 more positional argument (reportCallIssue)
  /Users/slott/Documents/Projects/ChatTUI/spike/bug_report.py:8:20 - error: Return type is unknown (reportUnknownVariableType)
3 errors, 0 warnings, 0 informations 


% pyright --version
pyright 1.1.351
```

**Comment by erictraut** [comment]
Pyright's behavior is correct here, so I don't consider this a bug. The error you're seeing is because of the definition of `MutableMapping.setdefault` found in the typeshed `typing.pyi` stub file.

```python
class MutableMapping(Mapping[_KT, _VT]):
    ...
    @overload
    def setdefault(self: MutableMapping[_KT, _T | None], __key: _KT, __default: None = None) -> _T | None: ...
    @overload
    def setdefault(self, __key: _KT, __default: _VT) -> _VT: ...
```

The first overload doesn't apply in this case because `dict[str, str]` is not compatible with `MutableMapping[_Kt, _T | None]`. That means the second overload is used, and it requires that the specified default is of type `_VT` (which is `str` in your example). You're attempting to pass a `str | None` value, which is a type violation.

If you think that the `MutableMapping` definition is incorrect, please file a bug with the [typeshed project](https://github.com/python/typeshed). The typeshed stubs are community-maintained and are used by all of the major Python type checkers.

**Comment by slott56** [comment]
Changing from `default: str | None = None`  to `default: str` can't work because then the  signature of the subclass `sefdefault()` doesn't match the parent class.

It's not clear how to preserve the `setdefault()` alternative overloads here.
